File: CsharpProjects/myLeetCode/Binary_Tree/Traversal mode of Binary tree/levelOrder/116. Populating Next Right Pointers in Each Node.cs
// 116. Populating Next Right Pointers in Each Node
// https://leetcode.com/problems/populating-next-right-pointers-in-each-node/

// 117. Populating Next Right Pointers in Each Node II
// https://leetcode.com/problems/populating-next-right-pointers-in-each-node-ii/

// 这两道题的代码完全一样。
#include <iostream>
#include <queue>

struct Node {
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(nullptr), right(nullptr), next(nullptr) {}
    Node(int x) : val(x), left(nullptr), right(nullptr), next(nullptr) {}
    Node(int x, Node* left, Node* right, Node* next)
        : val(x), left(left), right(right), next(next) {}
};

Node* connect(Node* root) {
    if (root == nullptr) return nullptr;

    std::queue<Node*> q;
    q.push(root);

    while (!q.empty()) {
        int size = q.size();
        Node* prev = nullptr;

        for (int i = 0; i < size; i++) {
            Node* curr = q.front();
            q.pop();

            if (prev) prev->next = curr;

            if (curr->left) q.push(curr->left);
            if (curr->right) q.push(curr->right);

            prev = curr;
        }
    }

    return root;
}

// Utility function to print the next pointers
void printNextPointers(Node* root) {
    Node* currLevel = root;

    while (currLevel != nullptr) {
        Node* curr = currLevel;

        while (curr != nullptr) {
            std::cout << curr->val << " -> ";
            curr = curr->next;
        }

        std::cout << "nullptr" << std::endl;
        currLevel = currLevel->left;
    }
}

int main() {
    // Create a sample binary tree
    Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    // root->right->left = new Node(6);
    root->right->right = new Node(7);

    Node* connectedRoot = connect(root);
    std::cout << "Next pointers:" << std::endl;
    printNextPointers(connectedRoot);

    // Clean up
    delete root->right->right;
    // delete root->right->left;
    delete root->left->right;
    delete root->left->left;
    delete root->right;
    delete root->left;
    delete root;

    return 0;
}

File: C++/代码随想录/Binary_Tree/Traversal mode of Binary tree/levelOrder/116. Populating Next Right Pointers in Each Node.py
# 116. Populating Next Right Pointers in Each Node
# https://leetcode.com/problems/populating-next-right-pointers-in-each-node/

# 117. Populating Next Right Pointers in Each Node II
# https://leetcode.com/problems/populating-next-right-pointers-in-each-node-ii/

# 这两道题的代码完全一样。
class Node:
    def __init__(self, val=0, left=None, right=None, next=None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return None

        leftmost = root
        while leftmost:
            dummy = Node(0)
            tail = dummy
            cur = leftmost

            while cur:
                if cur.left:
                    tail.next = cur.left
                    tail = tail.next
                if cur.right:
                    tail.next = cur.right
                    tail = tail.next
                cur = cur.next

            leftmost = dummy.next

        return root

# Test case
root = Node(1, Node(2, Node(4), Node(5)), Node(3, Node(6), Node(7)))
sol = Solution()
print(sol.connect(root))  # Connects the nodes as described in the problem statement
